pnet_packet/src/util.rs

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) 2014, 2015 Robert Clipsham <robert@octarineparrot.com>
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Utilities for working with packets, eg. checksumming
+
+use ip::IpNextHeaderProtocol;
+use pnet_macros_support::types::u16be;
+
+use std::net::{Ipv4Addr, Ipv6Addr};
+use std::slice;
+use std::u8;
+
+/// Convert value to byte array
+pub trait Octets {
+    /// Output type - bytes array
+    type Output;
+
+    /// Return value as bytes (big-endian order)
+    fn octets(&self) -> Self::Output;
+}
+
+impl Octets for u64 {
+    type Output = [u8; 8];
+
+    fn octets(&self) -> Self::Output {
+        [(*self >> 56) as u8,
+         (*self >> 48) as u8,
+         (*self >> 40) as u8,
+         (*self >> 32) as u8,
+         (*self >> 24) as u8,
+         (*self >> 16) as u8,
+         (*self >> 8) as u8,
+         *self as u8]
+    }
+}
+
+impl Octets for u32 {
+    type Output = [u8; 4];
+
+    fn octets(&self) -> Self::Output {
+        [(*self >> 24) as u8, (*self >> 16) as u8, (*self >> 8) as u8, *self as u8]
+    }
+}
+
+impl Octets for u16 {
+    type Output = [u8; 2];
+
+    fn octets(&self) -> Self::Output {
+        [(*self >> 8) as u8, *self as u8]
+    }
+}
+
+impl Octets for u8 {
+    type Output = [u8; 1];
+
+    fn octets(&self) -> Self::Output {
+        [*self]
+    }
+}
+
+/// Calculates a checksum. Used by ipv4 and icmp. The two bytes starting at `skipword * 2` will be
+/// ignored. Supposed to be the checksum field, which is regarded as zero during calculation.
+pub fn checksum(data: &[u8], skipword: usize) -> u16be {
```

Comment:
It would be nice if we could have the checksum functions somewhere easy to use without pulling in `pnet_packet` maybe? The only dependency they have on this crate is `IpNextHeaderProtocol`, which we could easily swap out for the corresponding primitive.

Reply:
I don't mean it as a the movement of this should block this PR. Just want to bring up the topic and see what you think. `pnet_base` maybe?

Reply:
Perhaps, but that's for another pull request.

I think if you're writing your own packet parsing code you can probably handle writing checksum functions - I'm not sure that it's worth having *another* crate for this. Maybe it is *shruggs*.

Reply:
I didn't mean creating yet another crate. I agree that would be overkill. I just meant that if we agreed that one crate we already have would be suitable then it could be there. But yes, it can stay here as well. I'll likely write my own for my packet crate anyway.

Reply:
I think, given that you're likely to write your own, so would anyone else. I think it should stay in pnet_packet, unless a motivating reason to do otherwise appears.